jax/experimental/slab/slab.py

```diff
@@ -0,0 +1,367 @@
+
+def make_allocating_op(op, ref_op):
+  def made_op(slab, *xs: SlabView):
+    out_sds = jax.eval_shape(
+        ref_op, *[jax.ShapeDtypeStruct(x.shape, x.dtype) for x in xs])
+    slab, out = slab_alloc(slab, out_sds.shape, out_sds.dtype)
+    slab = op(slab, xs, out)
+    return slab, out
+  return made_op
+
+add = make_allocating_op(partial(elementwise, jax.lax.add), jax.lax.add)
+mul = make_allocating_op(partial(elementwise, jax.lax.mul), jax.lax.mul)
+tanh = make_allocating_op(partial(elementwise, jax.lax.tanh), jax.lax.tanh)
+# TODO(frostig,mattjj): eval_shape fails
+#matmul = make_allocating_op(_matmul, lambda a, b: a @ b)
+def matmul(slab, av, bv):
+  out_shape = (av.shape[0], bv.shape[1])
+  slab, out = slab_alloc(slab, out_shape, av.dtype)
+  slab = _matmul(slab, (av, bv), out)
+  return slab, out
+
+def parse_arr(i, s):
+  shape = eval(s)
+  return np.random.RandomState(i).normal(size=shape).astype(np.float32)
+
+def print_seg(msg):
+  print()
+  print(f'-- {msg}')
+  print()
+
+def make_jaxpr_slab_write(slab, view, inval):
+  return jax.make_jaxpr(
+      lambda slab, x: slab_write(slab, view, (0, 0), x))(slab, inval)
+
+def make_jaxpr_slab_read(slab, view, outval_shape):
+  return jax.make_jaxpr(
+      lambda slab: slab_read(slab, view, (0, 0), outval_shape))(slab)
+
+def slab_download(slab, v):
+  if not static_shape(v.shape): raise Exception
+  return slab_read(slab, v, (0,) * v.ndim(), v.shape)
+
+def slab_upload(slab, x):
+  slab, xv = slab_alloc(slab, x.shape, x.dtype)
+  slab = slab_write(slab, xv, (0,) * x.ndim, x)
+  return slab, xv
+
+def chain(slab, fs, *argss, unary=False):
+  if callable(fs):
+    fs = [fs] * len(argss)
+  outss = []
+  for f, args in zip(fs, argss):
+    if unary:
+      slab, outs = f(slab, args)
+    else:
+      slab, outs = f(slab, *args)
+    outss.append(outs)
+  return slab, outss
+
+def test_binop(op, ref_op, slab, x, y):
+  z = ref_op(x, y)
+  slab, xv = slab_upload(slab, x)
+  slab, yv = slab_upload(slab, y)
+  slab, zv = op(slab, xv, yv)
+  assert jnp.allclose(slab_download(slab, xv), x, atol=1e-4)
+  assert jnp.allclose(slab_download(slab, yv), y, atol=1e-4)
+  assert jnp.allclose(slab_download(slab, zv), z, atol=1e-4)
+
+def main(args):
+  xs = map(parse_arr, range(len(args)), args)
+  assert all(len(x.shape) == 2 for x in xs)
+
+  slab = slab_make(1024)
+
+  x, y, *_ = xs
+  test_binop(add, jax.lax.add, slab, x, x)
+  test_binop(mul, jax.lax.mul, slab, x, x)
+  test_binop(matmul, lambda a, b: a @ b, slab, x, y)
+
+  def put(slab, x):
+    slab, v = slab_upload(slab, x)
+    print_seg('slab_read result')
+    print(slab_download(slab, v))
+    return slab, v
+
+  slab, vals = chain(slab, put, *xs, unary=True)
+
+  if len(vals) >= 2:
+    x, y, *_ = vals
+    slab, z = mul(slab, x, x)
+    print_seg('mul')
+    print(slab_download(slab, z))
+    slab, w = add(slab, x, z)
+    print_seg('add')
+    print(slab_download(slab, w))
+
+
+if __name__ == '__main__':
+  main(sys.argv[1:])
```

Comment:
Should we just add a new test file?

Reply:
Yeah, I should factor out into tests in both modules.